source/lac/trilinos_vector.cc

```diff
@@ -199,18 +199,11 @@ namespace TrilinosWrappers
 
     void
     Vector::reinit (const Epetra_Map &input_map,
-                    const bool        omit_zeroing_entries)
+                    const bool)
     {
       nonlocal_vector.reset();
 
-      if (vector->Map().SameAs(input_map)==false)
-        vector.reset (new Epetra_FEVector(input_map));
-      else if (omit_zeroing_entries == false)
-        {
-          const int ierr = vector->PutScalar(0.);
-          (void)ierr;
-          Assert (ierr == 0, ExcTrilinosError(ierr));
-        }
+      vector.reset (new Epetra_FEVector(input_map));
```

Comment:
Although, thinking about it: this deletes the old vector, which may use a stale communicator. What your experiment shows is that the destructor of `Epetra_FEVector` does not do any MPI communication. That makes sense, but it could also make sense to put, say, an `MPI_Barrier` or other synchronisation into the destructor. So I guess we get lucky.